File: lua/test/key_result_entity_test.lua
-- KeyResult entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("shortcut_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("KeyResultEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:KeyResult(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = key_result_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"update", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "key_result." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set SHORTCUT_TEST_KEY_RESULT_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local key_result_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.key_result")))
    local key_result_ref01_data = nil
    if #key_result_ref01_data_raw > 0 then
      key_result_ref01_data = helpers.to_map(key_result_ref01_data_raw[1][2])
    end

    -- UPDATE
    local key_result_ref01_ent = client:KeyResult(nil)
    local key_result_ref01_data_up0_up = {
      id = key_result_ref01_data["id"],
    }

    local key_result_ref01_markdef_up0_name = "name"
    local key_result_ref01_markdef_up0_value = "Mark01-key_result_ref01_" .. tostring(setup.now)
    key_result_ref01_data_up0_up[key_result_ref01_markdef_up0_name] = key_result_ref01_markdef_up0_value

    local key_result_ref01_resdata_up0_result, err = key_result_ref01_ent:update(key_result_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local key_result_ref01_resdata_up0 = helpers.to_map(type(key_result_ref01_resdata_up0_result) == 'table' and key_result_ref01_resdata_up0_result.data_get and key_result_ref01_resdata_up0_result:data_get() or key_result_ref01_resdata_up0_result)
    assert.is_not_nil(key_result_ref01_resdata_up0)
    assert.are.equal(key_result_ref01_resdata_up0["id"], key_result_ref01_data_up0_up["id"])
    assert.are.equal(key_result_ref01_resdata_up0[key_result_ref01_markdef_up0_name], key_result_ref01_markdef_up0_value)

    -- LOAD
    local key_result_ref01_match_dt0 = {
      id = key_result_ref01_data["id"],
    }
    local key_result_ref01_data_dt0_loaded, err = key_result_ref01_ent:load(key_result_ref01_match_dt0, nil)
    assert.is_nil(err)
    local key_result_ref01_data_dt0_load_result = helpers.to_map(type(key_result_ref01_data_dt0_loaded) == 'table' and key_result_ref01_data_dt0_loaded.data_get and key_result_ref01_data_dt0_loaded:data_get() or key_result_ref01_data_dt0_loaded)
    assert.is_not_nil(key_result_ref01_data_dt0_load_result)
    assert.are.equal(key_result_ref01_data_dt0_load_result["id"], key_result_ref01_data["id"])

  end)
end)

function key_result_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/key_result/KeyResultTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read key_result test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "key_result01", "key_result02", "key_result03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("SHORTCUT_TEST_KEY_RESULT_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["SHORTCUT_TEST_KEY_RESULT_ENTID"] = idmap,
    ["SHORTCUT_TEST_LIVE"] = "FALSE",
    ["SHORTCUT_TEST_EXPLAIN"] = "FALSE",
    ["SHORTCUT_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["SHORTCUT_TEST_KEY_RESULT_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["SHORTCUT_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["SHORTCUT_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["SHORTCUT_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["SHORTCUT_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: rb/test/key_result_entity_test.rb
# KeyResult entity test

require "minitest/autorun"
require "json"
require_relative "../Shortcut_sdk"
require_relative "runner"

class KeyResultEntityTest < Minitest::Test
  def test_create_instance
    testsdk = ShortcutSDK.test(nil, nil)
    ent = testsdk.KeyResult(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = key_result_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["update", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "key_result." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set SHORTCUT_TEST_KEY_RESULT_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    key_result_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.key_result")))
    key_result_ref01_data = nil
    if key_result_ref01_data_raw.length > 0
      key_result_ref01_data = Helpers.to_map(key_result_ref01_data_raw[0][1])
    end

    # UPDATE
    key_result_ref01_ent = client.KeyResult(nil)
    key_result_ref01_data_up0_up = {
      "id" => key_result_ref01_data["id"],
    }

    key_result_ref01_markdef_up0_name = "name"
    key_result_ref01_markdef_up0_value = "Mark01-key_result_ref01_#{setup[:now]}"
    key_result_ref01_data_up0_up[key_result_ref01_markdef_up0_name] = key_result_ref01_markdef_up0_value

    key_result_ref01_resdata_up0_result = key_result_ref01_ent.update(key_result_ref01_data_up0_up, nil)
    key_result_ref01_resdata_up0 = Helpers.to_map(key_result_ref01_resdata_up0_result)
    assert !key_result_ref01_resdata_up0.nil?
    assert_equal key_result_ref01_resdata_up0["id"], key_result_ref01_data_up0_up["id"]
    assert_equal key_result_ref01_resdata_up0[key_result_ref01_markdef_up0_name], key_result_ref01_markdef_up0_value

    # LOAD
    key_result_ref01_match_dt0 = {
      "id" => key_result_ref01_data["id"],
    }
    key_result_ref01_data_dt0_loaded = key_result_ref01_ent.load(key_result_ref01_match_dt0, nil)
    key_result_ref01_data_dt0_load_result = Helpers.to_map(key_result_ref01_data_dt0_loaded)
    assert !key_result_ref01_data_dt0_load_result.nil?
    assert_equal key_result_ref01_data_dt0_load_result["id"], key_result_ref01_data["id"]

  end
end

def key_result_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "key_result", "KeyResultTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = ShortcutSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["key_result01", "key_result02", "key_result03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["SHORTCUT_TEST_KEY_RESULT_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "SHORTCUT_TEST_KEY_RESULT_ENTID" => idmap,
    "SHORTCUT_TEST_LIVE" => "FALSE",
    "SHORTCUT_TEST_EXPLAIN" => "FALSE",
    "SHORTCUT_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["SHORTCUT_TEST_KEY_RESULT_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["SHORTCUT_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["SHORTCUT_APIKEY"],
      },
      extra || {},
    ])
    client = ShortcutSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["SHORTCUT_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["SHORTCUT_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
